Unittester: add `on_new_connection` + `on_load` + `skip_tests` options

carlopi opened the pull request:
This PR adds on top of https://github.com/duckdb/duckdb/pull/17972 the logic needed to actually run end-to-end the test suite while running custom logic on init / on connection init.
In particular 3 new fields:
* `on_new_connection` for SQL statements to be executed on creating a new connection (such as `USE db;` that is not persistent cross connections)
* `on_load` that instruct how to handle sqllogic level `load` statements. Currently accepts only parameters `` (default) or `skip`, but could be expanded in the future to allow other modes
* `skip_tests`, a JSON array of tests that have to be skipped. This allow to encode, within a config, the tests that are expected to fail / or irrelevant for the configuration.

Those plus the preexisting infrastructure allows to run end-to-end tests with different configurations, such as the included config with an encrypted database. Running sqlite or ducklake or else as default persistent backend is also possible, but left as exercise.

Added configs are checked in at `test/configs/`, and there is a CI job that execute them.
I think after initial iterations, proper CI setup would be that only added / modified configs are checked on PRs, while a scheduled job can then check older configs are still in order.

Contributions by:
* @lnkuiper that did a speed run to change `ParseJSONMap`/`ParseJSON` that now allow also parsing [nested] arrays, thanks a lot!!
* @Mytherin for feedback
* @ccfelius for testing of initial stage, I took liberty to include a config with encrypted databases inspired by that

carlopi committed:
Implement on_new_connection

carlopi committed:
Add 2 configuration to source tree, persistent_db and encrypted_db

carlopi committed:
@lnkuiper's patch to handle also [nested] arrays in ParseJSONMap

carlopi committed:
config files to json array

carlopi committed:
Test configs in CI

carlopi committed:
Fix code quality

carlopi committed:
OnConnect -> OnInit, and add on_load, on_new_connection and skip_tests

carlopi committed:
Move from iteration on VARCHAR to unordered_set<string>

carlopi committed:
Concatenate OnInit and OnConnection

carlopi committed:
ComplexJSONType doesn't need serialization

carlopi committed:
Add relevant include

carlopi committed:
Fixup CI, adding back dependencies

carlopi commented:
I should have added `"skip_compiled": "true",` to both configs, that would cut some unnecessary work.

carlopi commented:
I think this is ready, point with more chances to have unintended consequences to the codebase is the JSON parsing part, but there it should still be positive that we can now parse JSON array (nested at any depth and possibly containing objects).

I did sent #18054 that relies on this and it's on top, so this makes more sense to be reviewed first.

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments below

carlopi committed:
Move workflow from LinuxRelease to Main, and run only changed configurations

carlopi committed:
CI for changed workflows: skip removed ones

carlopi committed:
remove persistent_db, basically a copy of --force-storage

Mytherin commented on .github/workflows/Main.yml:
```diff
@@ -516,3 +516,40 @@ jobs:
+    - name: Test changed configurations
+      shell: bash
+      run: |
+        git diff --name-only ${BASE_BRANCH} test/configs| xargs -I{} echo $./build/release/test/unittest --test-config {}
```
Can you explain the reasoning behind adding the git diff here?

carlopi committed:
Rework slighly extensions statically linked in

carlopi committed:
Move --force-storage --force-restart to configuration force_storage_restart.json

carlopi committed:
Rework slighly extensions statically linked in

carlopi committed:
Move to --force-storage specific run to configuration force_storage.json

carlopi committed:
Add in DEBUG_STACKTRACE, always handy

carlopi committed:
Absorb also LATEST_STORAGE:1 via latest_storage.json

carlopi committed:
Format yml / refactor yml

carlopi committed:
Eat up also --verify_vector tests in the config ones

carlopi committed:
Run tests as long as build succeded

carlopi committed:
Absorb PREFETCH_ALL_PARQUET_FILES into prefetch_all_parquet_files.json

carlopi committed:
[CI] if: always()

carlopi committed:
Actually add prefetch_all_parquet_files

carlopi committed:
Fix syntax + avoid BLOCK_VERIFICATION

carlopi committed:
reldebug -> release

carlopi committed:
Fixup verify-vector=

carlopi committed:
fix prefetch

carlopi committed:
More iterations

carlopi committed:
no inet

carlopi committed:
Avoid triggering workflows on changes to test/configs

carlopi committed:
fix prefetch_all_parquet

carlopi committed:
Fix deps

carlopi committed:
Yet another iteration

carlopi committed:
Remove [DUCKDB_]PREFETCH_ALL_PARQUET_FILES

Reccomented workaround is like --test-config test/configs/prefetch_all_parquet_files.json

carlopi committed:
Skip a test with prefetch_all_parquet_files

Mytherin commented on test/configs/prefetch_all_parquet_files.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "description": "Debug setting around testing of parquet files",
+  "skip_compiled": "true",
+  "on_init": "INSTALL parquet FROM 'build/release/repository'; LOAD parquet; SET prefetch_all_parquet_files=true;",
```
Maybe we can add a new setting to avoid needing to do this, e.g.:

`"extensions": ["parquet"]`

We already have `core_functions` that we always load in `SQLLogicTestRunner::LoadDatabase`

carlopi committed:
[Unittester] Add  for extensions to be loaded at startup

carlopi committed:
Format fix

carlopi commented:
One relevant bit: this is NOT an exact match of current testing setup, for this to be properly a superset the following extra changes would be needed:
* Add back in `PRAGMA verify_fetch_row` (currently that will show some errors though, to be fixed on a side), see comment at test/configs/force_storage.json
* Add back BLOCK_VERIFICATION as compile time setting (that will shows some issues though, left a comment https://github.com/duckdb/duckdb/pull/18042/files#diff-bff2b2408261ca028d311f52dbfe908fc5eac03e62f3388b968a102ee6190146R343, to be added back once those are solved)
* Add back both `fts` and `inet`, currently there are some issues on how to resolve path for extensions in `skip_tests` / `{TEST_BASE_NAME}`

My judgement is that exactly the same test surface is not required, and getting this started is more relevant, but I though it's relevant to write this down, and will open relevant issues to track those problems.

Mytherin commented:
Thanks!
